rules_bazel/java/java_integration_test.bzl

```diff
@@ -0,0 +1,82 @@
+def _compare_with_goldens_test_impl(ctx):
+    diff_output = ctx.outputs.diff_output
+    check_diff_script = ctx.outputs.check_diff_script
+    input = ctx.attr.api_target
+    srcs = ctx.files.srcs
+    api_name = ctx.attr.name
+    
+    script = """
+    mkdir codegen_tmp
+    unzip -j {input} -d codegen_tmp
+    unzip -j {input_resource_name} -d  codegen_tmp
+    unzip -j {input_test} -d codegen_tmp
+    cd codegen_tmp
+    # Remove unneeded non-Java files, like MANIFEST
+    rm -rf $(find . -type f ! -name "*.java")
+    cd ..
+    diff codegen_tmp test/integration/goldens/{api_name}/ > {diff_output}
+    """.format(
+        diff_output = diff_output.path,
+        input = input[0][JavaInfo].source_jars[0].path,
+        input_resource_name = input[1][JavaInfo].source_jars[0].path,
+        input_test = input[2][JavaInfo].source_jars[0].path,
+        api_name = api_name
+    )
+    ctx.actions.run_shell(
+        inputs = srcs + [
+            input[0][JavaInfo].source_jars[0], 
+            input[1][JavaInfo].source_jars[0], 
+            input[2][JavaInfo].source_jars[0],
+        ],
+        outputs = [diff_output],
+        command = script,
+    )
+
+
+    check_diff_script_content = """
+    cd ${{BUILD_WORKSPACE_DIRECTORY}}
+    if [ -s {diff_output} ]
+    then
+        cat {diff_output}
```

Comment:
Is it a final variant or a debugging artifacts, which will be removed before submitting?

Reply:
This is a final variant, `diff_output` is a text file that contains all the diffs between goldens and codegen, here we are checking if the `diff_output` is empty. Empty means no difference then test passes, but if it;s not empty, `exit 1` will help us throw error, and `cat {diff_output}` prints the diff information to the generated `sh` file.